### DISCUSSION TITLE
Best way of stepping thru pages and check current step?

### DISCUSSION THREAD

**Discussion by simpel** [vercel/next.js#51640] [App Router] [answered]
Hi all!

Using the app dir and server components, what's the best way of handling a user flow? We're making a quiz that the user can click thru by answering questions. For each question we have a page (quiz/1, quiz/2 etc.). We need to track what question the user is on and if she tries to visit a page she has not answered yet we'd like to redirect her back to the current question.

But how to solve this using server components. Normally I would make a state (using zustand) and then check if the step visited is larger than the last asked question. If it is just redirect. But I can't keep state on the server, so how would you do it? Is there a good way of keeping server session? or should I just accept that I need to make some client components?

**Comment by icyJoseph** [comment, marked as answer]
Maybe make a `/quiz/[step]/page.tsx` kind of thing? 

```tsx
// quiz/[step]/page.tsx

import { redirect } from "next/navigation";
import { cookies } from "next/headers";
import { revalidatePath } from "next/cache";

export default function QuizStep({ params }: { params: { step: string } }) {
  const stepAsNumber = Number(params.step);

  if (Number.isNaN(stepAsNumber) || stepAsNumber < 1) redirect("/quiz/1");

  const pending = cookies().get("pending-question")?.value;

  const pendingAsNumber = Number(pending);

  if (Number.isNaN(pendingAsNumber)) redirect("/quiz/reset");

  if (!Number.isInteger(pendingAsNumber) || !Number.isInteger(stepAsNumber))
    redirect("/quiz/reset");

  // now we have "valid" pending step and requested step

  // trying to skip ahead
  if (stepAsNumber > pendingAsNumber) redirect(`/quiz/${pendingAsNumber}`);

  // send them to see their answer to this question
  if (stepAsNumber < pendingAsNumber) redirect(`/answers/${stepAsNumber}`);

  async function answerQuestion() {
    "use server";
    cookies().set({
      name: "pending-question",
      value: `${stepAsNumber + 1}`,
      httpOnly: true,
      path: "/",
    });
    revalidatePath("/quiz");
    redirect(`/quiz/${stepAsNumber + 1}`);
  }

  // now we know we are at the required number
  return (
    <div>
      <h2>Quiz: {stepAsNumber}</h2>

      <form action={answerQuestion}>
        <input type="text" />
        <button>Answer</button>
      </form>
    </div>
  );
}
```

The reset route would be something like:

```ts
// quiz/reset/route.ts
import { NextRequest, NextResponse } from "next/server";

export const GET = (req: NextRequest) => {
  req.nextUrl.pathname = "/quiz/1";
  const response = NextResponse.redirect(req.nextUrl);

  response.cookies.set({
    name: "pending-question",
    value: "1",
    httpOnly: true,
    path: "/",
  });

  return response;
};
```

It seems to work well :)  - don't forget to activate server actions

**Reply by icyJoseph** [reply]
A build/start run:

https://github.com/vercel/next.js/assets/21013447/2a5ac5f6-d722-4403-8e69-0a5af9e6168e

Though the counter there is not part of the page, but the layout... 🤔  so maybe state that lives in the page itself gets destroyed, who knows, ran out of time to try these!

**Reply by simpel** [reply]
Nice Ice @icyJoseph! So basically use cookies and use server ... I guess thats a valid way of solving my issue! Thanks!! You're awesome!
